File: blue-core/src/main/java/blue/time/TempoMap.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2000-2006, 2023, 2025 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.time;

import blue.components.lines.Line;
import blue.components.lines.LinePoint;
import blue.utility.XMLUtilities;
import electric.xml.Element;
import electric.xml.Elements;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.StringTokenizer;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Manages tempo changes across a musical timeline.
 * 
 * TempoMap maintains an ordered list of TempoPoints, where each point defines
 * a tempo at a specific beat position. The tempo between points can either:
 * - Stay constant until the next point (CONSTANT curve type)
 * - Interpolate linearly to the next point (LINEAR curve type)
 * 
 * Entries are always kept sorted by beat position.
 * 
 * When enabled is false, the tempo map behaves as if there's a single
 * constant tempo of 60 BPM (the default).
 *
 * @author Steven Yi
 */
public class TempoMap {
    
    /** Default tempo when tempo map is disabled or empty */
    public static final double DEFAULT_TEMPO = 60.0;
    
    /**
     * Listener interface for TempoMap changes.
     */
    public interface TempoMapListener {
        void tempoMapChanged();
    }

    private final List<TempoPoint> points = new ArrayList<>();
    private final List<TempoMapListener> listeners = new CopyOnWriteArrayList<>();
    private final List<PropertyChangeListener> propertyListeners = new CopyOnWriteArrayList<>();
    
    /** Whether tempo changes are enabled (Use Tempo checkbox) */
    private boolean enabled = false;
    
    /** Whether the tempo editor is expanded/visible in the UI */
    private boolean visible = false;

    public TempoMap() {
        points.add(new TempoPoint(0, DEFAULT_TEMPO, CurveType.LINEAR));
    }

    public TempoMap(TempoMap tempoMap) {
        for (TempoPoint point : tempoMap.points) {
            points.add(new TempoPoint(point));
        }
        this.enabled = tempoMap.enabled;
        this.visible = tempoMap.visible;
    }
    
    // ========== Enabled/Visible Properties ==========
    
    /**
     * Returns whether tempo changes are enabled.
     * When disabled, beatsToSeconds uses a constant 60 BPM.
     */
    public boolean isEnabled() {
        return enabled;
    }
    
    /**
     * Sets whether tempo changes are enabled.
     */
    public void setEnabled(boolean enabled) {
        if (this.enabled != enabled) {
            boolean oldValue = this.enabled;
            this.enabled = enabled;
            firePropertyChange("enabled", oldValue, enabled);
            fireChanged();
        }
    }
    
    /**
     * Returns whether the tempo editor is visible/expanded in the UI.
     */
    public boolean isVisible() {
        return visible;
    }
    
    /**
     * Sets whether the tempo editor is visible/expanded in the UI.
     */
    public void setVisible(boolean visible) {
        if (this.visible != visible) {
            boolean oldValue = this.visible;
            this.visible = visible;
            firePropertyChange("visible", oldValue, visible);
        }
    }
    
    // ========== Listener Management ==========
    
    public void addListener(TempoMapListener listener) {
        if (listener != null && !listeners.contains(listener)) {
            listeners.add(listener);
        }
    }
    
    public void removeListener(TempoMapListener listener) {
        listeners.remove(listener);
    }
    
    public void addPropertyChangeListener(PropertyChangeListener listener) {
        if (listener != null && !propertyListeners.contains(listener)) {
            propertyListeners.add(listener);
        }
    }
    
    public void removePropertyChangeListener(PropertyChangeListener listener) {
        propertyListeners.remove(listener);
    }
    
    private void fireChanged() {
        recalculateAccumulatedTimes();
        for (TempoMapListener listener : listeners) {
            listener.tempoMapChanged();
        }
    }
    
    private void firePropertyChange(String propertyName, Object oldValue, Object newValue) {
        PropertyChangeEvent event = new PropertyChangeEvent(this, propertyName, oldValue, newValue);
        for (PropertyChangeListener listener : propertyListeners) {
            listener.propertyChange(event);
        }
    }
    
    // ========== List Operations ==========
    
    public int size() {
        return points.size();
    }
    
    public boolean isEmpty() {
        return points.isEmpty();
    }
    
    /**
     * Gets the tempo point at the given index.
     */
    public TempoPoint getPoint(int index) {
        return points.get(index);
    }
    
    /**
     * Gets an unmodifiable view of all tempo points.
     */
    public List<TempoPoint> getPoints() {
        return Collections.unmodifiableList(points);
    }
    
    /**
     * Gets the beat position at the given index.
     */
    public double getBeat(int index) {
        return points.get(index).getBeat();
    }
    
    /**
     * Gets the tempo at the given index.
     */
    public double getTempo(int index) {
        return points.get(index).getTempo();
    }
    
    /**
     * Gets the curve type at the given index.
     */
    public CurveType getCurveType(int index) {
        return points.get(index).getCurveType();
    }
    
    /**
     * Adds a tempo point. The list is automatically sorted by beat.
     */
    public void addTempoPoint(TempoPoint point) {
        points.add(new TempoPoint(point));
        sortEntries();
        fireChanged();
    }
    
    /**
     * Updates the tempo point at the given index.
     * 
     * @param index the index to update
     * @param beat the new beat position (must be >= 0)
     * @param tempo the new tempo in BPM (must be > 0)
     * @throws IllegalArgumentException if beat < 0 or tempo <= 0
     */
    public void setTempoPoint(int index, double beat, double tempo) {
        setTempoPoint(index, beat, tempo, points.get(index).getCurveType());
    }
    
    /**
     * Updates the tempo point at the given index.
     * 
     * @param index the index to update
     * @param beat the new beat position (must be >= 0)
     * @param tempo the new tempo in BPM (must be > 0)
     * @param curveType the curve type for transition to next point
     * @throws IllegalArgumentException if beat < 0 or tempo <= 0
     */
    public void setTempoPoint(int index, double beat, double tempo, CurveType curveType) {
        TempoPoint point = points.get(index);
        point.setBeat(beat);
        point.setTempo(tempo);
        point.setCurveType(curveType);
        sortEntries();
        fireChanged();
    }
    
    /**
     * Sets only the curve type at the given index.
     */
    public void setCurveType(int index, CurveType curveType) {
        points.get(index).setCurveType(curveType);
        fireChanged();
    }
    
    /**
     * Removes the tempo point at the given index.
     * Note: Cannot remove the last tempo point.
     * 
     * @param index the index to remove
     * @throws IllegalStateException if trying to remove the last tempo point
     */
    public void removeTempoPoint(int index) {
        if (points.size() <= 1) {
            throw new IllegalStateException("Cannot remove the last tempo point");
        }
        points.remove(index);
        fireChanged();
    }
    
    /**
     * Clears all tempo points and resets to default (beat 0, tempo 60, LINEAR).
     */
    public void reset() {
        points.clear();
        points.add(new TempoPoint(0, DEFAULT_TEMPO, CurveType.LINEAR));
        fireChanged();
    }
    
    private void sortEntries() {
        Collections.sort(points);
    }
    
    /**
     * Recalculates the cached beat positions for all tempo points using the provided MeterMap.
     * This should be called when the MeterMap changes, as MeasureBeatsTime positions
     * depend on the meter map for conversion to beats.
     * 
     * @param meterMap the meter map for measure-to-beat conversion
     */
    public void recalculateBeatPositions(MeterMap meterMap) {
        for (TempoPoint point : points) {
            point.recalculateBeat(meterMap);
        }
        sortEntries();
        recalculateAccumulatedTimes();
        fireChanged();
    }
    
    /**
     * Recalculates the accumulated time for each tempo point.
     * Must be called after any modification to the tempo map.
     */
    private void recalculateAccumulatedTimes() {
        if (points.isEmpty()) return;
        
        points.get(0).accumulatedTime = 0;
        
        for (int i = 1; i < points.size(); i++) {
            TempoPoint last = points.get(i - 1);
            TempoPoint curr = points.get(i);
            
            double deltaBeat = curr.getBeat() - last.getBeat();
            
            if (deltaBeat == 0) {
                curr.accumulatedTime = last.accumulatedTime;
            } else if (last.getCurveType() == CurveType.CONSTANT) {
                // Constant tempo: simple calculation
                double factor = 60.0 / last.getTempo();
                curr.accumulatedTime = last.accumulatedTime + (factor * deltaBeat);
            } else {
                // Linear interpolation: use area under curve
                double factor1 = 60.0 / last.getTempo();
                double factor2 = 60.0 / curr.getTempo();
                double acceleration = (factor2 - factor1) / deltaBeat;
                curr.accumulatedTime = last.accumulatedTime + 
                    getAreaUnderCurve(factor1, deltaBeat, acceleration);
            }
        }
    }

    /**
     * Creates a TempoMap from a time warp string (legacy format).
     * Format: "beat1 tempo1 beat2 tempo2 ..."
     * All segments use LINEAR interpolation.
     * 
     * @param timeWarpString the time warp string
     * @return the created TempoMap, or null if parsing fails
     */
    public static TempoMap createTempoMap(String timeWarpString) {
        TempoMap tm = new TempoMap();
        tm.points.clear(); // Remove default point

        StringTokenizer st = new StringTokenizer(timeWarpString);

        if (st.countTokens() % 2 != 0) {
            // not an even amount of tokens!
            return null;
        }

        while (st.hasMoreTokens()) {
            try {
                String timeStr = st.nextToken();
                String tempoStr = st.nextToken();

                double beat = Double.parseDouble(timeStr);
                double tempo = Double.parseDouble(tempoStr);

                if (beat < 0.0 || tempo <= 0.0) {
                    return null;
                }

                tm.points.add(new TempoPoint(beat, tempo, CurveType.LINEAR));
            } catch (Exception e) {
                // if there's any errors whatsoever, return null
                // and let the calling procedure handle it
                return null;
            }
        }
        
        if (tm.points.isEmpty()) {
            return null;
        }
        
        // When created from a string, the tempo map should be enabled
        // (this is used by TimeWarpProcessor for explicit time warping)
        tm.enabled = true;
        tm.recalculateAccumulatedTimes();
        return tm;
    }
    
    /**
     * Gets the instantaneous tempo at the given beat position.
     * 
     * @param beat the beat position
     * @return the tempo in BPM at that beat
     */
    public double getTempoAt(double beat) {
        if (!enabled || points.isEmpty()) {
            return DEFAULT_TEMPO;
        }
        
        if (beat <= 0.0) {
            return points.get(0).getTempo();
        }
        
        // Find the segment containing this beat
        for (int i = 0; i < points.size() - 1; i++) {
            TempoPoint curr = points.get(i);
            TempoPoint next = points.get(i + 1);
            
            if (beat >= curr.getBeat() && beat < next.getBeat()) {
                if (curr.getCurveType() == CurveType.CONSTANT) {
                    return curr.getTempo();
                } else {
                    // Linear interpolation
                    double t = (beat - curr.getBeat()) / (next.getBeat() - curr.getBeat());
                    return curr.getTempo() + t * (next.getTempo() - curr.getTempo());
                }
            }
        }
        
        // Beyond last point - use last tempo
        return points.get(points.size() - 1).getTempo();
    }

    /**
     * Converts beats to seconds using the tempo map.
     * If tempo is disabled, uses constant 60 BPM.
     */
    public double beatsToSeconds(double beat) {
        if (beat == 0.0) {
            return 0.0;
        }
        
        // If disabled, use constant default tempo
        if (!enabled) {
            return beat * (60.0 / DEFAULT_TEMPO);
        }

        for (int i = 0; i < points.size() - 1; i++) {
            TempoPoint curr = points.get(i);
            TempoPoint next = points.get(i + 1);

            if (beat >= curr.getBeat() && beat < next.getBeat()) {
                double deltaBeat = beat - curr.getBeat();
                
                if (curr.getCurveType() == CurveType.CONSTANT) {
                    // Constant tempo in this segment
                    double factor = 60.0 / curr.getTempo();
                    return curr.accumulatedTime + (factor * deltaBeat);
                } else {
                    // Linear interpolation
                    double factor1 = 60.0 / curr.getTempo();
                    double factor2 = 60.0 / next.getTempo();
                    double segmentLength = next.getBeat() - curr.getBeat();
                    
                    double acceleration = (factor2 - factor1) / segmentLength;
                    double t = getAreaUnderCurve(factor1, deltaBeat, acceleration);
                    
                    return curr.accumulatedTime + t;
                }
            }
        }

        // Beyond last point - use constant tempo from last point
        TempoPoint lastPoint = points.get(points.size() - 1);
        double factor = 60.0 / lastPoint.getTempo();
        double deltaBeat = beat - lastPoint.getBeat();

        return lastPoint.accumulatedTime + (factor * deltaBeat);
    }

    private static double getAreaUnderCurve(double factor1, double deltaBeat,
            double acceleration) {
        return (factor1 * deltaBeat)
                + (0.5 * acceleration * Math.pow(deltaBeat, 2));
    }

    /**
     * Converts seconds to beats using the tempo map.
     * If tempo is disabled, uses constant 60 BPM.
     */
    public double secondsToBeats(double seconds) {
        if (seconds == 0.0) {
            return 0.0;
        }
        
        // If disabled, use constant default tempo
        if (!enabled) {
            return seconds * (DEFAULT_TEMPO / 60.0);
        }

        if (points.size() == 1) {
            double factor = points.get(0).getTempo() / 60.0;
            return seconds * factor;
        }

        for (int i = 0; i < points.size() - 1; i++) {
            TempoPoint curr = points.get(i);
            TempoPoint next = points.get(i + 1);

            if (seconds < next.accumulatedTime) {
                double beat0 = curr.getBeat();
                double time0 = curr.accumulatedTime;
                
                if (curr.getCurveType() == CurveType.CONSTANT) {
                    // Constant tempo: simple calculation
                    double factor = curr.getTempo() / 60.0;
                    double elapsedTime = seconds - time0;
                    return beat0 + (elapsedTime * factor);
                } else {
                    // Linear interpolation
                    // BASED ON CODE BY ISTVAN VARGA Csound Mailing List - April 13, 2006
                    double beat1 = next.getBeat();
                    double btime0 = 60.0 / curr.getTempo();
                    double btime1 = 60.0 / next.getTempo();

                    double x;
                    if (btime0 == btime1) {
                        double elapsedTime = seconds - time0;
                        x = elapsedTime / btime0;
                    } else {
                        double a = 0.5 * (btime1 - btime0) / (beat1 - beat0);
                        double b = btime0;
                        double c = time0 - seconds;
                        x = (Math.sqrt(b * b - (4 * a * c)) - b) / (2 * a);
                    }

                    return x + beat0;
                }
            }
        }

        // Beyond last point
        TempoPoint last = points.get(points.size() - 1);
        double factor = last.getTempo() / 60.0;
        return ((seconds - last.accumulatedTime) * factor) + last.getBeat();
    }

    @Override
    public String toString() {
        StringBuilder buffer = new StringBuilder();
        buffer.append("[TempoMap enabled=").append(enabled).append("]\n");

        for (TempoPoint point : points) {
            buffer.append(point.getBeat())
                  .append(" : ")
                  .append(point.getTempo())
                  .append(" (")
                  .append(point.getCurveType())
                  .append(")\n");
        }

        return buffer.toString();
    }
    
    // ========== XML Serialization ==========
    
    /**
     * Save TempoMap to XML.
     */
    public Element saveAsXML() {
        Element retVal = new Element("tempoMap");
        
        retVal.addElement(XMLUtilities.writeBoolean("enabled", enabled));
        retVal.addElement(XMLUtilities.writeBoolean("visible", visible));
        
        for (TempoPoint point : points) {
            retVal.addElement(point.saveAsXML());
        }
        
        return retVal;
    }
    
    /**
     * Load TempoMap from XML.
     * Supports both new format (tempoPoint elements) and legacy format (beatTempoPair elements).
     */
    public static TempoMap loadFromXML(Element data) throws Exception {
        TempoMap tm = new TempoMap();
        tm.points.clear();
        
        Elements children = data.getElements();
        while (children.hasMoreElements()) {
            Element child = children.next();
            String name = child.getName();
            
            switch (name) {
                case "enabled" -> tm.enabled = Boolean.parseBoolean(child.getTextString());
                case "visible" -> tm.visible = Boolean.parseBoolean(child.getTextString());
                case "tempoPoint" -> tm.points.add(TempoPoint.loadFromXML(child));
                case "beatTempoPair" -> tm.points.add(TempoPoint.loadFromLegacyXML(child));
            }
        }
        
        // Ensure at least one point exists
        if (tm.points.isEmpty()) {
            tm.points.add(new TempoPoint(0, DEFAULT_TEMPO, CurveType.LINEAR));
        }
        
        tm.recalculateAccumulatedTimes();
        return tm;
    }
    
    // ========== Migration from Legacy Tempo ==========
    
    /**
     * Creates a TempoMap from legacy &lt;tempo&gt; XML element.
     * This is used for migrating old project files that used the
     * blue.score.tempo.Tempo class with a Line for tempo automation.
     * 
     * The Line's points (x=beat, y=tempo) are converted to TempoPoints.
     * The old system always used linear interpolation between points.
     * 
     * @param tempoElement the legacy &lt;tempo&gt; XML element
     * @return a new TempoMap with the same tempo data
     */
    public static TempoMap loadFromLegacyTempoXML(Element tempoElement) {
        TempoMap tm = new TempoMap();
        tm.points.clear();
        
        Elements nodes = tempoElement.getElements();
        Line line = null;
        
        while (nodes.hasMoreElements()) {
            Element node = nodes.next();
            switch (node.getName()) {
                case "enabled" -> tm.enabled = Boolean.parseBoolean(node.getTextString());
                case "visible" -> tm.visible = Boolean.parseBoolean(node.getTextString());
                case "line" -> line = Line.loadFromXML(node);
            }
        }
        
        // Convert Line points to TempoPoints
        if (line != null) {
            for (int i = 0; i < line.size(); i++) {
                LinePoint lp = line.getLinePoint(i);
                double beat = lp.getX();
                double tempoValue = lp.getY();
                
                // Old system always used linear interpolation
                tm.points.add(new TempoPoint(beat, tempoValue, CurveType.LINEAR));
            }
        }
        
        // Ensure at least one point exists
        if (tm.points.isEmpty()) {
            tm.points.add(new TempoPoint(0, DEFAULT_TEMPO, CurveType.LINEAR));
        }
        
        tm.recalculateAccumulatedTimes();
        return tm;
    }
}

File: packages/blue-data/src/time/tempo-map.ts
/**
 * TempoMap — manages tempo changes over time.
 * Mirrors the Java TempoMap class.
 *
 * Supports multiple TempoPoint entries with CONSTANT or LINEAR curves.
 * When disabled, always returns 60 BPM regardless of tempo points.
 */
import { CurveType, parseCurveType } from './curve-type';
import { TempoPoint } from './tempo-point';
import { Element } from '../serialization/xml-reader';
import { TimePosition } from './time-position';
import { TimeContext } from './time-context';

export class TempoMap {
  private points: TempoPoint[] = [new TempoPoint(undefined, 60, CurveType.CONSTANT)];
  private _enabled = false;
  private _visible = false;
  private listeners: Array<() => void> = [];

  /** Copy constructor. */
  constructor(source?: TempoMap) {
    if (source) {
      this._enabled = source._enabled;
      this._visible = source._visible;
      this.points = source.points.map((p) => {
        const np = new TempoPoint(p.position, p.tempo, p.curveType);
        np.enabled = p.enabled;
        np.visible = p.visible;
        np.beat = p.beat;
        np.accumulatedTime = p.accumulatedTime;
        return np;
      });
    }
  }

  // ─── Accessors ───

  /** Get tempo. With no args returns first point's tempo; with index returns that point's tempo. */
  getTempo(index?: number): number {
    if (index !== undefined) return this.points[index]?.tempo ?? 60;
    return this.points[0]?.tempo ?? 60;
  }

  setTempo(bpm: number): void {
    if (this.points.length === 1) {
      this.points[0].tempo = bpm;
      this.recalculateAccumulatedTimes();
    }
  }

  /** Get the beat position at a specific index. */
  getBeat(index: number): number {
    return this.points[index]?.beat ?? 0;
  }

  /** Get the curve type at a specific index. */
  getCurveType(index: number): CurveType {
    return this.points[index]?.curveType ?? CurveType.CONSTANT;
  }

  /** Get the tempo point at a specific index. Alias for getTempoPoint. */
  getPoint(index: number): TempoPoint {
    return this.points[index];
  }

  isEnabled(): boolean {
    return this._enabled;
  }
  setEnabled(enabled: boolean): void {
    this._enabled = enabled;
    this.recalculateAccumulatedTimes();
    this.fireListeners();
  }

  isVisible(): boolean {
    return this._visible;
  }
  setVisible(visible: boolean): void {
    if (this._visible !== visible) {
      this._visible = visible;
      this.fireListeners();
    }
  }

  size(): number {
    return this.points.length;
  }

  getTempoPoints(): ReadonlyArray<TempoPoint> {
    return this.points;
  }

  getTempoPoint(index: number): TempoPoint {
    return this.points[index];
  }

  /** Add a tempo point, optionally with context for bar-based position resolution. */
  addTempoPoint(point: TempoPoint, context?: TimeContext): void {
    this.points.push(point);
    if (context) {
      this.recalculateBeatPositions(context);
    } else {
      this.recalculateAccumulatedTimes();
    }
    this.fireListeners();
  }

  /**
   * Set tempo point at index.
   * Overloads:
   * - setTempoPoint(index, point: TempoPoint)
   * - setTempoPoint(index, beat: number, tempo: number, curveType?: CurveType)
   * - setTempoPoint(index, position: TimePosition, tempo: number, curveType: CurveType, context: TimeContext)
   */
  setTempoPoint(index: number, ...args: any[]): void {
    if (args[0] instanceof TempoPoint) {
      this.points[index] = args[0];
      this.recalculateAccumulatedTimes();
    } else if (args[0] instanceof TimePosition) {
      const [position, tempo, curveType, context] = args;
      this.points[index] = new TempoPoint(position, tempo, curveType);
      this.recalculateBeatPositions(context);
    } else if (typeof args[0] === 'number') {
      const beat = args[0];
      const tempo = args[1];
      const curveType = args[2] ?? CurveType.CONSTANT;
      const point = new TempoPoint(beat, tempo, curveType);
      this.points[index] = point;
      this.recalculateAccumulatedTimes();
    }
    this.fireListeners();
  }

  removeTempoPoint(index: number): void {
    if (this.points.length <= 1) {
      throw new Error('Cannot remove the last tempo point');
    }
    this.points.splice(index, 1);
    this.recalculateAccumulatedTimes();
    this.fireListeners();
  }

  /** Reset to single default point (60 BPM at beat 0). */
  reset(): void {
    this.points = [new TempoPoint(undefined, 60, CurveType.CONSTANT)];
    this.recalculateAccumulatedTimes();
    this.fireListeners();
  }

  /** Replace all data from a source TempoMap. */
  replaceAll(source: TempoMap): void {
    this._enabled = source._enabled;
    this._visible = source._visible;
    this.points = source.points.map((p) => {
      const np = new TempoPoint(p.position, p.tempo, p.curveType);
      np.enabled = p.enabled;
      np.visible = p.visible;
      np.beat = p.beat;
      np.accumulatedTime = p.accumulatedTime;
      return np;
    });
    this.recalculateAccumulatedTimes();
    this.fireListeners();
  }

  addListener(listener: () => void): void {
    this.listeners.push(listener);
  }

  private fireListeners(): void {
    for (const listener of this.listeners) {
      listener();
    }
  }

  // ─── Conversion ───

  getBeatDuration(): number {
    return 60.0 / this.getTempo();
  }

  /**
   * Get the tempo at a given beat position.
   * For CONSTANT curves, returns the tempo of the segment containing the beat.
   * For LINEAR curves, interpolates between points.
   */
  getTempoAt(beat: number): number {
    if (!this._enabled) return 60;

    // Find the segment containing this beat
    for (let i = this.points.length - 1; i >= 0; i--) {
      if (beat >= this.points[i].beat) {
        const point = this.points[i];
        // Beyond last point: constant tempo
        if (i >= this.points.length - 1) return point.tempo;
        // CONSTANT: use this point's tempo
        if (point.curveType === CurveType.CONSTANT) return point.tempo;
        // LINEAR: interpolate
        const nextPoint = this.points[i + 1];
        const segmentLength = nextPoint.beat - point.beat;
        if (segmentLength <= 0) return point.tempo;
        const fraction = (beat - point.beat) / segmentLength;
        return point.tempo + fraction * (nextPoint.tempo - point.tempo);
      }
    }
    return this.points[0].tempo;
  }

  /**
   * Convert beats to seconds.
   * For disabled maps: 1 beat = 1 second (60 BPM).
   * For enabled maps: accumulates time across segments with CONSTANT or LINEAR curves.
   */
  beatsToSeconds(beats: number): number {
    if (!this._enabled) return beats; // 60 BPM → 1 beat = 1 second

    // Find the segment containing this beat
    let segIndex = 0;
    for (let i = this.points.length - 1; i >= 0; i--) {
      if (beats >= this.points[i].beat) {
        segIndex = i;
        break;
      }
    }

    const point = this.points[segIndex];
    const accumulatedTime = point.accumulatedTime;

    // Beyond last point: extrapolate with constant tempo
    if (segIndex >= this.points.length - 1) {
      const deltaBeats = beats - point.beat;
      return accumulatedTime + deltaBeats * (60.0 / point.tempo);
    }

    const deltaBeats = beats - point.beat;
    const nextPoint = this.points[segIndex + 1];

    if (point.curveType === CurveType.CONSTANT) {
      return accumulatedTime + deltaBeats * (60.0 / point.tempo);
    }

    // LINEAR: compute area under curve
    // tempo(t) = t0 + (t1 - t0) * (x / deltaBeat)
    // integral = 60 * [ x / t(x) ] from 0 to deltaBeats
    // Using the formula: time = 60 * 2 * deltaBeats / (t0 + t1) for linear interpolation
    const t0 = point.tempo;
    const t1 = nextPoint.tempo;
    const segmentBeats = nextPoint.beat - point.beat;

    if (t0 === t1) {
      return accumulatedTime + deltaBeats * (60.0 / t0);
    }

    // Quadratic formula based on Istvan Varga's solution (April 2006)
    const factor1 = 60.0 / t0;
    const acceleration = (60.0 / t1 - factor1) / segmentBeats;
    return accumulatedTime + factor1 * deltaBeats + 0.5 * acceleration * deltaBeats * deltaBeats;
  }

  /**
   * Convert seconds to beats.
   * Reverse of beatsToSeconds.
   */
  secondsToBeats(seconds: number): number {
    if (!this._enabled) return seconds; // 60 BPM → 1 second = 1 beat

    // Find the segment containing this time
    let segIndex = 0;
    for (let i = this.points.length - 1; i >= 0; i--) {
      if (seconds >= this.points[i].accumulatedTime) {
        segIndex = i;
        break;
      }
    }

    const point = this.points[segIndex];
    const elapsed = seconds - point.accumulatedTime;

    // Beyond last point: extrapolate with constant tempo
    if (segIndex >= this.points.length - 1) {
      return point.beat + elapsed * (point.tempo / 60.0);
    }

    const nextPoint = this.points[segIndex + 1];

    if (point.curveType === CurveType.CONSTANT) {
      return point.beat + elapsed * (point.tempo / 60.0);
    }

    // LINEAR: quadratic formula
    const t0 = point.tempo;
    const t1 = nextPoint.tempo;
    const segmentBeats = nextPoint.beat - point.beat;

    if (t0 === t1) {
      return point.beat + elapsed * (t0 / 60.0);
    }

    const factor1 = 60.0 / t0;
    const acceleration = (60.0 / t1 - factor1) / segmentBeats;
    // time = factor1 * x + 0.5 * acceleration * x^2
    // Solve: 0.5 * a * x^2 + factor1 * x - elapsed = 0
    // x = (sqrt(factor1^2 + 2*a*elapsed) - factor1) / a
    const discriminant = factor1 * factor1 + 2 * acceleration * elapsed;
    if (acceleration === 0) {
      return point.beat + elapsed / factor1;
    }
    return point.beat + (Math.sqrt(Math.max(0, discriminant)) - factor1) / acceleration;
  }

  /**
   * Recompute cached beat positions and accumulated times for all points.
   */
  recalculateAccumulatedTimes(): void {
    if (this.points.length === 0) return;

    for (const point of this.points) {
      if (point.position.isBeatTime()) {
        point.beat = point.position.getValue();
      }
    }

    if (this.points.every((point) => point.position.isBeatTime())) {
      this.points.sort((a, b) => a.beat - b.beat);
    }

    this.points[0].accumulatedTime = 0;

    for (let i = 1; i < this.points.length; i++) {
      const prev = this.points[i - 1];
      const cur = this.points[i];

      // Only recompute beat from position if position is beat-based
      // (bar-based positions are resolved by recalculateBeatPositions)
      if (cur.position.isBeatTime()) {
        cur.beat = cur.position.getValue();
      }

      const deltaBeats = cur.beat - prev.beat;

      if (prev.curveType === CurveType.CONSTANT) {
        cur.accumulatedTime = prev.accumulatedTime + deltaBeats * (60.0 / prev.tempo);
      } else {
        // LINEAR
        const factor1 = 60.0 / prev.tempo;
        const acceleration = deltaBeats > 0 ? (60.0 / cur.tempo - factor1) / deltaBeats : 0;
        cur.accumulatedTime =
          prev.accumulatedTime +
          factor1 * deltaBeats +
          0.5 * acceleration * deltaBeats * deltaBeats;
      }
    }
  }

  /**
   * Recalculate beat positions using a TimeContext for bar-based positions.
   * Points with bar-based positions (BBT, BBST, BBF) are converted to beats
   * using the context's MeterMap before recalculating accumulated times.
   */
  recalculateBeatPositions(context: TimeContext): void {
    if (this.points.length === 0) return;

    for (const point of this.points) {
      if (point.position && !point.position.isBeatTime()) {
        // Bar-based position: convert to beats
        point.beat = point.position.toBeats(context);
      }
    }

    this.points.sort((a, b) => a.beat - b.beat);

    this.recalculateAccumulatedTimes();
  }

  /**
   * Parse a legacy tempo map string format: "0 60 4 120 8 90"
   * Each pair is beat-position, tempo-BPM.
   */
  static createTempoMap(text: string): TempoMap | null {
    const tokens = text.trim().split(/\s+/);
    if (tokens.length < 2) return null;
    if (tokens.length % 2 !== 0) return null;

    const map = new TempoMap();
    map.points = [];
    map._enabled = true;

    for (let i = 0; i < tokens.length; i += 2) {
      const beat = parseFloat(tokens[i]);
      const tempo = parseFloat(tokens[i + 1]);

      if (isNaN(beat) || isNaN(tempo)) return null;
      if (beat < 0) return null;
      if (tempo <= 0) return null;

      const point = new TempoPoint(beat, tempo, CurveType.LINEAR);
      map.points.push(point);
    }

    map.recalculateAccumulatedTimes();
    return map;
  }

  equals(other: TempoMap): boolean {
    if (this._enabled !== other._enabled) return false;
    if (this.points.length !== other.points.length) return false;
    for (let i = 0; i < this.points.length; i++) {
      if (this.points[i].tempo !== other.points[i].tempo) return false;
      if (this.points[i].beat !== other.points[i].beat) return false;
      if (this.points[i].curveType !== other.points[i].curveType) return false;
    }
    return true;
  }

  // ─── XML Serialization ───

  saveAsXML(): Element {
    const elem = new Element('tempoMap');
    elem.addElement('enabled').setText(this._enabled.toString());
    elem.addElement('visible').setText(this._visible.toString());

    for (const point of this.points) {
      elem.addElement(point.saveAsXML());
    }

    return elem;
  }

  static loadFromXML(data: Element): TempoMap {
    const map = new TempoMap();

    const enabled = data.getTextString('enabled');
    if (enabled !== null) map._enabled = enabled !== 'false';

    const visible = data.getTextString('visible');
    if (visible !== null) map._visible = visible === 'true';

    const points = data.getElements('tempoPoint');
    if (points.hasMoreElements()) {
      map.points = [];
      while (points.hasMoreElements()) {
        const point = TempoPoint.loadFromXML(points.next());
        map.points.push(point);
      }
    }

    // Also check legacy beatTempoPair format (only if no tempoPoint elements were found)
    const legacyPoints = data.getElements('beatTempoPair');
    const hasTempoPoints =
      map.points.length !== 1 || map.points[0].tempo !== 60 || map.points[0].beat !== 0;
    if (legacyPoints.hasMoreElements() && !hasTempoPoints) {
      map.points = [];
      while (legacyPoints.hasMoreElements()) {
        const pairElem = legacyPoints.next();
        const beat = parseFloat(pairElem.getAttribute('beat') ?? '0');
        const tempo = parseFloat(pairElem.getAttribute('tempo') ?? '60');
        const point = new TempoPoint(beat, tempo, CurveType.LINEAR);
        map.points.push(point);
      }
    }

    if (map.points.length === 0) {
      map.points = [new TempoPoint(undefined, 60, CurveType.CONSTANT)];
    }

    map.recalculateAccumulatedTimes();
    return map;
  }
}
